pubtools/_pulp/tasks/common.py

```diff
@@ -179,7 +182,11 @@ def publish_with_cache_flush(
         # hook implementation(s) may now flush pulp-derived caches and datastores
         pm.hook.task_pulp_flush()
 
-        # flush CDN cache
+        # if publish is being skipped then cdn flush should also be skipped
+        if "publish" in self.args.skip and "flush_cdn" not in self.args.skip:
```

Comment:
Minor complaint: I think this kind of logic would be better moved to where the arguments are initialized, e.g. in PulpTask.args method.

The reason is that you're adding a side-effect to publish_with_cache_flush here: the "args" may now be different after the method returns.  It's not easy for other developers to be aware that this is happening.  If someone writes code elsewhere which checks whether self.args.skip contains "flush_cdn", it'll behave differently depending on whether that code runs before or after this method, making it bug-prone.

Whereas if you moved this initialization into PulpTask.args, then the value is always consistent.  Everything else should only read from `args` and never write to it, making it easier to understand.

Reply:
@rohanpm  i've been thinking about what you suggested here but is it possible if the arguments haven't yet been parsed within PulpTask? (unless im mistaken)

prior to making any of these changes i had originally thought of making the change within step.py in the StepDecorator class (which was abandoned fairly early)

i believe it was something similar to
```
    def should_skip(self, instance):
        skip = (getattr(instance.args, "skip", None) or "").split(",")
        if "publish" in skip and not "flush_cdn" in skip:
            skip.append("flush_cdn")
        return self.machine_name in skip
```

but i guess above doesn't make sense because each step is independent of each other aside from having the same skip arguments
Maybe something like
```
    def should_skip(self, instance):
        skip = getattr(instance.args, "skip", None) or ""
        if self.machine_name == "flush_cdn" and "publish" in skip and "flush_cdn" not in skip:
            skip.append("flush_cdn")
        return self.machine_name in skip
```

i ran above through the tests and it seems to be passing the tests, i suppose we could maintain some kind of conditional_skip function where we could hold these and apply it within should_skip function
it ended up looking like:
```
    def conditional_skip(self, skip):
        cond_skip = skip
        if self.machine_name == "flush_cdn" and "publish" in cond_skip and "flush_cdn" not in cond_skip:
            cond_skip.append("flush_cdn")
        return cond_skip

    def should_skip(self, instance):
        skip = getattr(instance.args, "skip", None) or ""
        skip = self.conditional_skip(skip)
        return self.machine_name in skip
```
it seems to check off even if code is written elsewhere if im not mistaken and also other code is not reliant on publish_with_cache_flush so i suppose we could undo a ton of things we've done and still include the skip without changing code elsewhere

Reply:
out of curiosity i went ahead and cloned down latest master and applied above, the tests passed successfully (unexpected) i expected it to fail on baselines because flush_cdn should be skipped in addition to publish changing the baselines but i guess i was wrong?

Reply:
i pushed up the corresponding changes related to what i'm referring to above on top of the latest master, let me know what you think

thanks

Reply:
> is it possible if the arguments haven't yet been parsed within PulpTask? 

I'm not quite sure what you mean by this; `self.args` is a property using lazy initialization, the args are going to be parsed the first time they're requested (and never parsed again after that). The suggestion of moving some of the arg handling logic into that property would have no impact on when they get initialized, so if this is not a problem with the initialization right now then it wouldn't be a problem after making that change either.